### PR TITLE
Constraints in Lie group

### PR DESCRIPTION
**Pull request by florent-lamiraux**
Solver now handle constraints in Lie groups.
typedef RelativeTransformationR3xSO3 compute the relative position between two frames as an element of R3xSO(3).

### CODE REVIEW

File: include/hpp/constraints/fwd.hh [hunk 1]
```diff
@@ -145,7 +145,7 @@ namespace hpp {
     typedef GenericTransformation< RelativeBit | PositionBit                  > RelativePosition;
     typedef GenericTransformation< RelativeBit |               OrientationBit > RelativeOrientation;
     typedef GenericTransformation<               PositionBit | OrientationBit | OutputSE3Bit > TransformationSE3;
-    typedef GenericTransformation< RelativeBit | PositionBit | OrientationBit | OutputSE3Bit > RelativeTransformationSE3;
+    typedef GenericTransformation< RelativeBit | PositionBit | OrientationBit | OutputSE3Bit > RelativeTransformationR3xSO3;
```
Comment (jmirabel) [inline include/hpp/constraints/fwd.hh:148]: Shouldn't we rename the following too ?
- TransformationSE3,
- OutputSE3Bit

File: include/hpp/constraints/matrix-view.hh [hunk 2]
```diff
@@ -678,6 +682,16 @@ namespace Eigen {
...
+    template<class Archive>
+    void serialize(Archive & ar, const unsigned int version)
+    {
+      (void) version;
+      ar & BOOST_SERIALIZATION_NVP(m_nbRows);
+      ar & BOOST_SERIALIZATION_NVP(m_nbCols);
+      ar & BOOST_SERIALIZATION_NVP(m_rows);
+      ar & BOOST_SERIALIZATION_NVP(m_cols);
+    }
+    friend class boost::serialization::access;
```
Comment (jmirabel) [inline include/hpp/constraints/matrix-view.hh:694]: Serialization of MatrixBlock is already available in `hpp/constraints/serialization.hh`.
```suggestion
```

File: src/implicit.cc [hunk 3]
```diff
@@ -240,6 +228,8 @@ namespace hpp {
         parameterSize_ = computeParameterSize (comparison_);
       ar & BOOST_SERIALIZATION_NVP(function_);
       ar & BOOST_SERIALIZATION_NVP(rhsFunction_);
+      ar & BOOST_SERIALIZATION_NVP(activeRows_);
+      ar & BOOST_SERIALIZATION_NVP(inactiveRows_);
```
Comment (jmirabel) [inline src/implicit.cc:232]: For compactness of the archive,
```suggestion
      if (!Archive::is_saving::value)
        computeActiveRows();
```

File: src/solver/by-substitution.cc [hunk 4]
```diff
@@ -210,10 +210,11 @@ namespace hpp {
               array().cast<bool>();
             active = adpF.any();
           }
-          if (active) // If at least one element of adp is true
-            rows.push_back (BlockIndices::segment_t
-                            (row, constraints [i]->function ().
-                             outputDerivativeSize()));
+          if (active){ // If at least one element of adp is true
+	    for (const segment_t s : constraints [i]->activeRows()) {
+	      rows.push_back(segment_t(s.first+row, s.second));
```
Comment (jmirabel) [inline src/solver/by-substitution.cc:215]: I think this should work and is more C++11:
```suggestion
	      rows.emplace_back(s.first+row, s.second);
```

File: src/solver/hierarchical-iterative.cc [hunk 5]
```diff
@@ -403,10 +400,10 @@ namespace hpp {
             (constraints [i]->function ().activeDerivativeParameters().
              matrix()).eval();
           if (adp.any()) // If at least one element of adp is true
-            rows.push_back (BlockIndices::segment_t
-                            (row, constraints [i]->function ().
-                             outputDerivativeSize()));
-          row += constraints [i]->function ().outputDerivativeSize();
+	    for (const segment_t s : constraints [i]->activeRows()) {
+	      rows.push_back(segment_t(s.first+offset, s.second));
```
Comment (jmirabel) [inline src/solver/hierarchical-iterative.cc:404]:
```suggestion
	      rows.emplace_back(s.first+offset, s.second);
```

File: include/hpp/constraints/fwd.hh (hunk 1 above)
Comment (florent-lamiraux) [reply include/hpp/constraints/fwd.hh:148]: I agree.

File: include/hpp/constraints/matrix-view.hh [hunk 6]
```diff
@@ -17,10 +17,12 @@
 #ifndef HPP_CONSTRAINTS_MATRIX_VIEW_HH
 #define HPP_CONSTRAINTS_MATRIX_VIEW_HH
 
+#include <boost/serialization/nvp.hpp>
 #include <Eigen/Core>
 #include <vector>
 #include <iostream>
 #include <hpp/util/indent.hh>
+#include <hpp/util/serialization-fwd.hh>
```
Comment (jmirabel) [inline include/hpp/constraints/matrix-view.hh:25]:
```suggestion
```

File: include/hpp/constraints/matrix-view.hh [hunk 7]
```diff
@@ -17,10 +17,12 @@
 #ifndef HPP_CONSTRAINTS_MATRIX_VIEW_HH
 #define HPP_CONSTRAINTS_MATRIX_VIEW_HH
 
+#include <boost/serialization/nvp.hpp>
```
Comment (jmirabel) [inline include/hpp/constraints/matrix-view.hh:20]:
```suggestion
```

File: include/hpp/constraints/matrix-view.hh [hunk 8]
```diff
@@ -143,6 +145,8 @@ namespace Eigen {
         template <typename In0_t, typename In1_t> empty_struct (In0_t, In1_t) {}
         static inline Index size() { return 1; }
         inline const Index& operator[](const Index& i) const { return i; }
+	template<class Archive>
+	inline void serialize(Archive &, const unsigned int){}
```
Comment (jmirabel) [inline include/hpp/constraints/matrix-view.hh:149]: Do you really need this ?

**Review by jmirabel — CHANGES_REQUESTED** [review]
The indentation is a mix of space and tabs. Could you try to homogenize it ?

**Pull request merged**